rpc/send_test.go

```diff
@@ -32,3 +38,249 @@ func TestInvalidAddrLength(t *testing.T) {
+		SendNextTimeout: 1 * time.Second,
+		Timeout:         1 * time.Second,
+	}
+	_, err := sendPing(opts, []net.Addr{s.Addr()}, rpcContext)
+	if err == nil {
+		t.Fatalf("Unexpected success")
+	}
+	retryErr, ok := err.(util.Retryable)
+	if !ok {
+		t.Fatalf("Unexpected error type: %v", err)
+	}
+	if !retryErr.CanRetry() {
+		t.Errorf("Expected retryable error: %v", retryErr)
+	}
+}
+
+// TestUnretryableError verifies that Send returns an unretryable
+// error when it hits a critical error.
+func TestUnretryableError(t *testing.T) {
+	rpcContext := createNewTestRPCContext(t)
+	s := createAndStartNewServer(rpcContext, t)
+
+	opts := Options{
+		N:               1,
+		Ordering:        OrderStable,
+		SendNextTimeout: 1 * time.Second,
+		Timeout:         1 * time.Second,
+	}
+	getArgs := func(addr net.Addr) interface{} {
+		return &proto.PingRequest{}
+	}
+	// Make getRetry return a non-proto value so that the proto
+	// integrity check fails.
+	getReply := func() interface{} {
+		return 0
+	}
+	_, err := Send(opts, "Heartbeat.Ping", []net.Addr{s.Addr()}, getArgs, getReply, rpcContext)
+	if err == nil {
+		t.Fatalf("Unexpected success")
+	}
+	retryErr, ok := err.(util.Retryable)
+	if !ok {
+		t.Fatalf("Unexpected error type: %v", err)
+	}
+	if retryErr.CanRetry() {
+		t.Errorf("Unexpected retryable error: %v", retryErr)
+	}
+}
+
+// TestComplexScenarios verifies various complex success/failure scenarios by
+// mocking sendOne.
+func TestComplexScenarios(t *testing.T) {
+	rpcContext := createNewTestRPCContext(t)
+
+	testCases := []struct {
+		numSevers                int
+		numRequests              int
+		numErrors                int
+		numRetryableErrors       int
+		success                  bool
+		isRetryableErrorExpected bool
+	}{
+		// --- Success senarios ---
+		{1, 1, 0, 0, true, false},
+		{5, 1, 0, 0, true, false},
+		{5, 5, 0, 0, true, false},
+		// There are some errors, but enough number of RPCs succeed.
+		{5, 1, 1, 0, true, false},
+		{5, 1, 4, 0, true, false},
+		{5, 3, 2, 0, true, false},
+
+		// --- Failure scenarios ---
+		// Too many requests.
+		{1, 5, 0, 0, false, false},
+		// All RPCs fail.
+		{5, 1, 5, 0, false, false},
+		// Some RPCs fail and we do not have enough remaining clients.
+		{5, 3, 3, 0, false, false},
+		// All RPCs fail, but some of the errors are retryable.
+		{5, 1, 5, 1, false, true},
+		{5, 3, 5, 3, false, true},
+		// Some RPCs fail, but we do have enough remaining clients and recoverable errors.
+		{5, 3, 3, 1, false, true},
+		{5, 3, 4, 2, false, true},
+	}
+	for i, test := range testCases {
+		var serverAddrs []net.Addr
+		for j := 0; j < test.numSevers; j++ {
+			s := createAndStartNewServer(rpcContext, t)
+			defer s.Close()
+			serverAddrs = append(serverAddrs, s.Addr())
+		}
+
+		opts := Options{
+			N:               test.numRequests,
+			Ordering:        OrderStable,
+			SendNextTimeout: 1 * time.Second,
+			Timeout:         1 * time.Second,
+		}
+		getArgs := func(addr net.Addr) interface{} {
```

Comment:
Yes, they are called. Changed `getReply` to return `interface{}(nil)`. `getArgs` needs to return `addr` so that `sendOneFn` can take a different action for each server. Added a comment.